<launch>

  <arg name="camera" default="frontleft"/>
  <arg name="stream_ns" default="stream_image"/>

  <group ns="rtabmap">

    <!-- RGBD synchronization -->
    <node pkg="nodelet" type="nodelet" name="rgbd_sync"
          args="standalone rtabmap_ros/rgbd_sync" output="screen">
      <remap from="depth/image"
             to="/spot/$(arg stream_ns)/$(arg camera)_depth_in_visual_frame/image"/>
      <remap from="rgb/image"
             to="/spot/$(arg stream_ns)/$(arg camera)_fisheye_image/image"/>
      <remap from="rgb/camera_info"
             to="/spot/$(arg stream_ns)/$(arg camera)_fisheye_image/camera_info"/>
      <remap from="rgbd_image" to="rgbd_image"/>
      <param name="approx_sync"  value="true"/>
    </node>

    <!-- rtabmap main node -->
    <node name="rtabmap" pkg="rtabmap_ros" type="rtabmap" output="screen" args="--delete_db_on_start">
      <param name="frame_id" type="string" value="body" />
      <param name="subscribe_depth" type="bool" value="false"/>
      <param name="subscribe_rgbd" type="bool" value="true"/>
      <param name="subscribe_scan" type="bool" value="false"/>

      <remap from="odom" to="/spot/odometry"/>
      <remap from="rgbd_image" to="rgbd_image"/>
      <param name="queue_size" type="int" value="10"/>

      <!--RTAB-Map's parameters -->
      <param name="RGBD/NeighborLinkRefining" type="string" value="true"/>
      <param name="RGBD/ProximityBySpace"     type="string" value="true"/>
      <param name="RGBD/AngularUpdate"        type="string" value="0.01"/>
      <param name="RGBD/LinearUpdate"         type="string" value="0.01"/>
      <param name="RGBD/OptimizeFromGraphEnd" type="string" value="false"/>
      <param name="Grid/FromDepth"            type="string" value="false"/> <!-- occupancy grid from lidar -->
      <param name="Reg/Force3DoF"             type="string" value="true"/>
      <param name="Reg/Strategy"              type="string" value="1"/> <!-- 1=ICP -->

      <!-- ICP parameters -->
      <param name="Icp/VoxelSize"                 type="string" value="0.05"/>
      <param name="Icp/MaxCorrespondenceDistance" type="string" value="0.1"/>
    </node>

  </group>
</launch>
